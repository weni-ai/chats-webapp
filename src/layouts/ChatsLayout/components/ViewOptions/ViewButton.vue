<template>
  <section class="view-button-content">
    <UnnnicButton
      size="large"
      type="tertiary"
      class="view-button"
      @click="handleClick"
    >
      <p class="view-button__text">{{ $t('view-option') }}</p>
      <UnnnicIcon
        data-testid="header-icon-expand"
        size="md"
        :icon="expandedMore ? 'expand_more' : 'expand_less'"
        scheme="neutral-darkest"
      />
    </UnnnicButton>
  </section>
</template>

<script setup>
import { defineProps } from 'vue';

defineProps({
  expandedMore: Boolean,
  handleClick: {
    type: Function,
    default: null,
  },
});
</script>

<style lang="scss" scoped>
.view-button-content {
  display: flex;
  align-items: center;

  :deep(.unnnic-button) {
    width: 100%;
    display: flex;
    justify-content: space-between;
    padding: $unnnic-spacing-sm;
    align-self: stretch;
    border-top: 1px solid $unnnic-color-neutral-soft;
    border-bottom: 1px solid $unnnic-color-neutral-soft;
    border-left: 1px solid $unnnic-color-neutral-soft;
    background-color: $unnnic-color-neutral-lightest;
  }

  :deep(.unnnic-button--tertiary) {
    background-color: $unnnic-color-neutral-lightest;
  }

  :deep(.unnnic-button--tertiary:active:enabled) {
    background-color: $unnnic-color-neutral-lightest;
  }

  :deep(.unnnic-button--tertiary:hover:enabled) {
    background-color: $unnnic-color-neutral-lightest;
    border-top: 1px solid $unnnic-color-neutral-cleanest;
    border-bottom: 1px solid $unnnic-color-neutral-cleanest;
    border-left: 1px solid $unnnic-color-neutral-cleanest;
  }
}

.view-button {
  :deep(.unnnic-button__label) {
    width: 100%;
    display: flex;
    justify-content: space-between;
  }

  &__text {
    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-style: normal;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
