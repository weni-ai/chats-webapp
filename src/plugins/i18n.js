import Vue from 'vue';
import VueI18n from 'vue-i18n';

// eslint-disable-next-line camelcase
import pt_br from '@/locales/pt_br.json';
import en from '@/locales/en.json';
import es from '@/locales/es.json';

import moment from 'moment';

import('moment/locale/pt-br');
import('moment/locale/es');

moment.locale('pt-br');

Vue.use(VueI18n);

const messages = {
  'pt-br': pt_br,
  en,
  es,
};

const i18n = new VueI18n({
  locale: 'pt-br',
  fallbackLocale: 'pt-br',
  messages,
  dateTimeFormats: {
    'pt-br': {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
    en: {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
    es: {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
  },
});

export default i18n;
