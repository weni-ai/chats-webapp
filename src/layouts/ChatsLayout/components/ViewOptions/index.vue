<template>
  <footer class="chats-layout-footer">
    <ViewButton
      v-if="!isOpen"
      :handleClick="openDrawer"
      :expandedMore="isOpen"
    />
  </footer>
  <Teleport to="body">
    <div
      v-if="isOpen"
      class="chats-layout-drawer__overlay"
      @click="closeDrawer"
    ></div>
    <div
      class="chats-layout-drawer"
      :class="{ 'chats-layout-drawer--open': isOpen }"
    >
      <ViewButton
        :handleClick="closeDrawer"
        :expandedMore="isOpen"
      />
      <section class="chats-layout-drawer__options">
        <section class="chats-layout-drawer__option">
          <UnnnicSwitch
            v-model="sound"
            data-testid="switch-sound"
            size="small"
            :textRight="$t('preferences.notifications.sound')"
            @update:model-value="changeSound"
          />
        </section>

        <section class="chats-layout-drawer__option">
          <UnnnicButton
            v-if="showFlowsTriggerButton"
            data-testid="show-flows-trigger"
            :text="$t('flows')"
            size="small"
            type="tertiary"
            iconLeft="send"
            @mousedown.prevent
            @click="openFlowsTrigger"
          />
        </section>
        <section class="chats-layout-drawer__option">
          <UnnnicButton
            data-testid="show-quick-messages"
            :text="$t('quick_messages.title')"
            iconLeft="bolt"
            type="tertiary"
            size="small"
            @mousedown.prevent
            @click="openQuickMessage"
          />
        </section>
        <section class="chats-layout-drawer__option">
          <UnnnicButton
            v-if="dashboard"
            data-testid="show-dashboard"
            text="Dashboard"
            iconLeft="bar_chart_4_bars"
            type="tertiary"
            size="small"
            @mousedown.prevent
            @click="navigate('dashboard.manager')"
          />
        </section>
        <section class="chats-layout-drawer__option">
          <UnnnicButton
            class="chats-layout-drawer__button"
            :text="$t('chats.see_history')"
            iconLeft="history"
            type="tertiary"
            size="small"
            @click="navigate('closed-rooms')"
          />
        </section>
      </section>
    </div>
  </Teleport>
</template>

<script setup>
import { ref, onBeforeMount, defineEmits } from 'vue';
import ViewButton from './ViewButton.vue';
import { PREFERENCES_SOUND } from '@/services/api/websocket/soundNotification.js';
import { useRouter } from 'vue-router';

defineProps({
  showFlowsTriggerButton: {
    type: Boolean,
    default: false,
  },
  dashboard: {
    type: Boolean,
    default: false,
  },
});

const emit = defineEmits(['open-flows-trigger', 'show-quick-messages']);

const { push } = useRouter();

const isOpen = ref(false);
const sound = ref(false);

const closeDrawer = () => {
  isOpen.value = false;
};

const openDrawer = () => {
  isOpen.value = true;
};

const changeSound = () => {
  localStorage.setItem(PREFERENCES_SOUND, sound.value ? 'yes' : 'no');
};

const openFlowsTrigger = () => {
  emit('open-flows-trigger');
};

const openQuickMessage = () => {
  emit('show-quick-messages');
};

const navigate = (name) => {
  push({
    name,
  });
};

onBeforeMount(() => {
  sound.value = (localStorage.getItem(PREFERENCES_SOUND) || 'yes') === 'yes';
});
</script>

<style lang="scss" scoped>
.chats-layout-drawer__overlay {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.5);
  z-index: 999;
}

.chats-layout-drawer {
  position: fixed;
  bottom: -(calc(100vw * (3 / 12)));
  width: calc(100vw * (3 / 12));
  transition: bottom 0.3s ease-in-out;
  z-index: 1000;

  border-top: 1px solid $unnnic-color-neutral-soft;
  border-bottom: 1px solid $unnnic-color-neutral-soft;
  border-left: 1px solid $unnnic-color-neutral-soft;
  background-color: $unnnic-color-neutral-lightest;

  &--open {
    bottom: 0;
    display: flex;
    flex-direction: column;
  }

  &__option {
    width: 100%;

    &:hover {
      background-color: $unnnic-color-neutral-light;
    }

    :deep(.unnnic-switch) {
      padding: $unnnic-spacing-nano + 1px $unnnic-spacing-sm;
      width: 100%;
    }

    :deep(.unnnic-switch__label__small) {
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-gt;
      font-style: normal;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }

    :deep(.unnnic-button--size-small) {
      padding: 0 $unnnic-spacing-sm;
      width: 100%;
      display: flex;
      align-items: center;
      justify-content: start;
    }

    :deep(.material-symbols-rounded.unnnic-icon-size--sm) {
      font-size: $unnnic-avatar-size-nano;
      color: $unnnic-color-neutral-cloudy;
    }

    :deep(.unnnic-button__label) {
      color: $unnnic-color-neutral-cloudy;
    }
  }
}

section.chats-layout .sidebar .chats-layout-footer {
  padding-right: 0;
}
</style>
