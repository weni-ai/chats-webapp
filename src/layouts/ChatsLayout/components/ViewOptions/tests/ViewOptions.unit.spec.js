import {
  beforeAll,
  afterAll,
  describe,
  expect,
  it,
  vi,
  beforeEach,
} from 'vitest';
import { mount, config } from '@vue/test-utils';
import ViewOptions from '@/layouts/ChatsLayout/components/ViewOptions/index.vue';
import ViewButton from '@/layouts/ChatsLayout/components/ViewOptions/ViewButton.vue';
import i18n from '@/plugins/i18n';

const mockPush = vi.fn();
vi.mock('vue-router', () => ({
  useRouter: () => ({
    push: mockPush,
  }),
}));

beforeAll(() => {
  config.global.plugins = config.global.plugins.filter(
    (plugin) => plugin !== i18n,
  );
});

afterAll(() => {
  if (!config.global.plugins.includes(i18n)) {
    config.global.plugins.push(i18n);
  }
});

describe('ViewOptions', () => {
  let wrapper;
  let localStorageMock;

  beforeEach(() => {
    localStorageMock = {
      storage: {},
      getItem: vi.fn((key) => localStorageMock.storage[key]),
      setItem: vi.fn((key, value) => {
        localStorageMock.storage[key] = value;
      }),
      clear: vi.fn(() => {
        localStorageMock.storage = {};
      }),
    };

    Object.defineProperty(window, 'localStorage', {
      value: localStorageMock,
    });
  });

  const createWrapper = (props = {}) => {
    return mount(ViewOptions, {
      props,
      global: {
        mocks: {
          $t: (key) => key,
        },
        stubs: {
          ViewButton: ViewButton,
          UnnnicSwitch: {
            template: '<input type="checkbox" data-testid="switch-sound" />',
          },
          UnnnicButton: {
            template: '<button @click="handleClick"><slot /></button>',
            props: ['handleClick', 'expandedMore'],
          },
          teleport: true,
        },
      },
    });
  };

  describe('Props', () => {
    it('should receive showFlowsTriggerButton as true', () => {
      wrapper = createWrapper({ showFlowsTriggerButton: true });
      expect(wrapper.props('showFlowsTriggerButton')).toBe(true);
    });

    it('should receive dashboard as false', () => {
      wrapper = createWrapper({ dashboard: false });
      expect(wrapper.props('dashboard')).toBe(false);
    });

    it('should receive isViewMode as true', () => {
      wrapper = createWrapper({ isViewMode: true });
      expect(wrapper.props('isViewMode')).toBe(true);
    });
  });

  describe('Rendering Content', () => {
    it('should render UnnnicSwitch when isViewMode is false', async () => {
      wrapper = createWrapper({ isViewMode: false });
      wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      expect(wrapper.find('[data-testid="switch-sound"]').exists()).toBe(true);
    });

    it('should not render UnnnicSwitch when isViewMode is true', () => {
      wrapper = createWrapper({ isViewMode: true });
      expect(wrapper.find('[data-testid="switch-sound"]').exists()).toBe(false);
    });

    it('should update localStorage when changeSound is called', async () => {
      wrapper = createWrapper();

      wrapper.vm.sound = true;
      await wrapper.vm.changeSound();
      expect(localStorageMock.setItem).toHaveBeenCalledWith(
        'chats_preferencesSound',
        'yes',
      );
      expect(localStorageMock.getItem('chats_preferencesSound')).toBe('yes');
      expect(wrapper.vm.sound).toBe(true);

      wrapper.vm.sound = false;
      await wrapper.vm.changeSound();
      expect(localStorageMock.setItem).toHaveBeenCalledWith(
        'chats_preferencesSound',
        'no',
      );
      expect(localStorageMock.getItem('chats_preferencesSound')).toBe('no');
      expect(wrapper.vm.sound).toBe(false);
    });

    it('should pass the correct value to UnnnicSwitch v-model', async () => {
      wrapper = createWrapper({ isViewMode: false });
      wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      expect(
        wrapper.find('[data-testid="switch-sound"]').attributes('modelvalue'),
      ).toBe('true');
    });

    it('should update sound when UnnnicSwitch is toggled', async () => {
      wrapper = createWrapper({ isViewMode: false, sound: true });
      wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      const switchComponent = wrapper.findComponent(
        '[data-testid="switch-sound"]',
      );

      await switchComponent.setValue(false);
      expect(wrapper.vm.sound).toBe(false);
    });
  });

  describe('Events', () => {
    it('should open drawer when ViewButton is clicked', async () => {
      wrapper = createWrapper();
      await wrapper
        .find('[data-testid="view-btn"]')
        .find('button')
        .trigger('click');
      await wrapper.vm.$nextTick();

      expect(wrapper.find('[data-testid="drawer"]').exists()).toBe(true);
    });

    it('should emit open-flows-trigger when openFlowsTrigger is called', async () => {
      wrapper = createWrapper();
      await wrapper.vm.openFlowsTrigger();
      expect(wrapper.emitted('open-flows-trigger')).toBeTruthy();
    });

    it('should emit show-quick-messages when openQuickMessage is called', async () => {
      wrapper = createWrapper();
      await wrapper.vm.openQuickMessage();
      expect(wrapper.emitted('show-quick-messages')).toBeTruthy();
    });
  });

  describe('Navigation', () => {
    it('should call router push when navigate is called', async () => {
      wrapper = createWrapper();
      await wrapper.vm.navigate('dashboard.manager');

      expect(mockPush).toHaveBeenCalledWith({
        name: 'dashboard.manager',
      });
    });
    it('should render dashboard button when dashboard is true', async () => {
      wrapper = createWrapper({ dashboard: true });
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      expect(wrapper.find('[data-testid="show-dashboard"]').exists()).toBe(
        true,
      );
    });

    it('should render dashboard button when isViewMode is false', async () => {
      wrapper = createWrapper({ isViewMode: false });
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      expect(wrapper.find('[data-testid="show-dashboard"]').exists()).toBe(
        true,
      );
    });

    it('should not render dashboard button when dashboard is false and isViewMode is true', () => {
      wrapper = createWrapper({ dashboard: false, isViewMode: true });
      expect(wrapper.find('[data-testid="show-dashboard"]').exists()).toBe(
        false,
      );
    });

    it('should call navigate when dashboard button is clicked', async () => {
      wrapper = createWrapper({ dashboard: true });
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      await wrapper.find('[data-testid="show-dashboard"]').trigger('click');
      expect(mockPush).toHaveBeenCalledWith({ name: 'dashboard.manager' });
    });

    it('should render the see history button', async () => {
      wrapper = createWrapper();
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      expect(wrapper.find('[data-testid="show-see_history"]').exists()).toBe(
        true,
      );
    });

    it('should call navigate when see history button is clicked', async () => {
      wrapper = createWrapper();
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      await wrapper.find('[data-testid="show-see_history"]').trigger('click');
      expect(mockPush).toHaveBeenCalledWith({ name: 'closed-rooms' });
    });
  });

  describe('Snapshot', () => {
    it('should match the snapshot', () => {
      wrapper = createWrapper();
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Button Event Handlers', () => {
    it('should prevent mousedown default on flows trigger button', async () => {
      wrapper = createWrapper({ showFlowsTriggerButton: true });
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      const button = wrapper.find('[data-testid="show-flows-trigger"]');
      const preventDefault = vi.fn();

      await button.trigger('mousedown', { preventDefault });
      expect(preventDefault).toHaveBeenCalled();
    });

    it('should prevent mousedown default on quick messages button', async () => {
      wrapper = createWrapper({ isViewMode: false });
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      const button = wrapper.find('[data-testid="show-quick-messages"]');
      const preventDefault = vi.fn();

      await button.trigger('mousedown', { preventDefault });
      expect(preventDefault).toHaveBeenCalled();
    });

    it('should prevent mousedown default on dashboard button', async () => {
      wrapper = createWrapper({ dashboard: true });
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      const button = wrapper.find('[data-testid="show-dashboard"]');
      const preventDefault = vi.fn();

      await button.trigger('mousedown', { preventDefault });
      expect(preventDefault).toHaveBeenCalled();
    });
  });

  describe('Sound Switch Complete Interactions', () => {
    it('should handle full sound switch lifecycle', async () => {
      wrapper = createWrapper({ isViewMode: false });
      await wrapper.vm.openDrawer();
      await wrapper.vm.$nextTick();
      await wrapper
        .findComponent('[data-testid="switch-sound"]')
        .setValue(false);
      expect(wrapper.vm.sound).toBe(false);

      const switchComponent = wrapper.findComponent(
        '[data-testid="switch-sound"]',
      );
      await switchComponent.setValue(true);
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.sound).toBe(true);
      expect(localStorageMock.setItem).toHaveBeenCalledWith(
        'chats_preferencesSound',
        'yes',
      );

      await switchComponent.trigger('update:model-value', true);
      expect(wrapper.vm.sound).toBe(true);
    });
  });

  describe('Lifecycle Hooks', () => {
    it('should set initial sound value from localStorage on mount', async () => {
      localStorageMock.storage['chats_preferencesSound'] = 'yes';
      wrapper = createWrapper();

      expect(wrapper.vm.sound).toBe(true);
    });

    it('should default sound to true if no localStorage value exists', async () => {
      localStorageMock.storage = {};
      wrapper = createWrapper();

      expect(wrapper.vm.sound).toBe(true);
    });

    it('should clean up click event listener on unmount', async () => {
      const removeEventListenerSpy = vi.spyOn(document, 'removeEventListener');
      wrapper = createWrapper();
      wrapper.unmount();

      expect(removeEventListenerSpy).toHaveBeenCalledWith(
        'click',
        expect.any(Function),
      );
    });
  });
});
