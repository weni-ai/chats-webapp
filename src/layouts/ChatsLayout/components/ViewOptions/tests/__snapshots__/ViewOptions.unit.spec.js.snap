// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ViewOptions > Snapshot > should match the snapshot 1`] = `
<div
  data-v-app=""
>
  
  <footer
    class="chats-layout-footer"
    data-v-2df0aa29=""
  >
    <section
      class="view-button-content"
      data-testid="view-btn"
      data-v-2df0aa29=""
      data-v-905ee5f6=""
    >
      <button
        class="view-button"
        data-testid="header-btn"
        data-v-905ee5f6=""
        size="large"
        type="submit"
      >
        
        <p
          class="view-button__text"
          data-testid="header-text"
          data-v-905ee5f6=""
        >
          view-option
        </p>
        <span
          class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md"
          data-testid="header-icon-expand"
          data-v-76ceda5d=""
          data-v-905ee5f6=""
          translate="no"
        >
          expand_less
        </span>
        
      </button>
    </section>
  </footer>
  <teleport-stub
    data-testid="teleport-body"
    data-v-2df0aa29=""
    to="body"
  >
    <!--v-if-->
    <section
      class="chats-layout-drawer"
      data-testid="drawer"
      data-v-2df0aa29=""
    >
      <section
        class="view-button-content"
        data-testid="view-btn-drawer"
        data-v-2df0aa29=""
        data-v-905ee5f6=""
      >
        <button
          class="view-button"
          data-testid="header-btn"
          data-v-905ee5f6=""
          size="large"
          type="submit"
        >
          
          <p
            class="view-button__text"
            data-testid="header-text"
            data-v-905ee5f6=""
          >
            view-option
          </p>
          <span
            class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md"
            data-testid="header-icon-expand"
            data-v-76ceda5d=""
            data-v-905ee5f6=""
            translate="no"
          >
            expand_less
          </span>
          
        </button>
      </section>
      <section
        class="chats-layout-drawer__options"
        data-v-2df0aa29=""
      >
        <section
          class="chats-layout-drawer__option"
          data-v-2df0aa29=""
        >
          <input
            data-testid="switch-sound"
            data-v-2df0aa29=""
            modelvalue="true"
            textright="preferences.notifications.sound"
            type="checkbox"
          />
        </section>
        <section
          class="chats-layout-drawer__option"
          data-v-2df0aa29=""
        >
          <button
            data-testid="show-flows-trigger"
            data-v-2df0aa29=""
            iconleft="send"
            size="small"
            text="flows"
            type="submit"
          >
            
            
          </button>
        </section>
        <section
          class="chats-layout-drawer__option"
          data-v-2df0aa29=""
        >
          <button
            data-testid="show-quick-messages"
            data-v-2df0aa29=""
            iconleft="bolt"
            size="small"
            text="quick_messages.title"
            type="submit"
          >
            
            
          </button>
        </section>
        <section
          class="chats-layout-drawer__option"
          data-v-2df0aa29=""
        >
          <button
            data-testid="show-dashboard"
            data-v-2df0aa29=""
            iconleft="bar_chart_4_bars"
            size="small"
            text="Dashboard"
            type="submit"
          >
            
            
          </button>
        </section>
        <section
          class="chats-layout-drawer__option"
          data-v-2df0aa29=""
        >
          <button
            class="chats-layout-drawer__button"
            data-testid="show-see_history"
            data-v-2df0aa29=""
            iconleft="history"
            size="small"
            text="chats.see_history"
            type="submit"
          >
            
            
          </button>
        </section>
      </section>
    </section>
  </teleport-stub>
  
</div>
`;
